276 move evaluate to server

TalalBazerbachi opened the pull request:
Evaluating models has been moved to the Server package, whenever a new RMSE value is in the database,  the UI gets updated.
Before you test this branch:
-you should run dump V16 first.
-run server for a few minutes.


This resolves [#276 ]

TalalBazerbachi committed:
#276 move evaluation to server and implement a runnable class to train models on multiple threas

TalalBazerbachi committed:
#276 update database when new model should be evaluated

TalalBazerbachi committed:
#276 update RMSE when new value is in the database

TalalBazerbachi committed:
Merge branch 'F10-Association_AType-Model' into 276_Move_Evaluate_To_Server

TalalBazerbachi committed:
#276 fix errors

TalalBazerbachi committed:
Merge branch 'develop' into 276_Move_Evaluate_To_Server

TalalBazerbachi committed:
#276 Update evaluation to comply with databse updates

TalalBazerbachi committed:
#276  fix error when inserting model strategy

TalalBazerbachi committed:
#276 update rmse value when new value is in database

TalalBazerbachi committed:
Merge branch 'develop' into 276_Move_Evaluate_To_Server

jeffbuonamici commented on JavaFXUserInterface/src/main/java/external/ModelDAOImpl.java:
```diff
@@ -172,6 +184,64 @@ public String getGetModelEvaluation(int modelID, String assetTypeID) {
     }
-}
\ No newline at end of file
+
+    @Override
+    public ModelStrategy getModelStrategy(int modelID, int assetTypeID) throws SQLException {
```
Add comments

TalalBazerbachi replied:
added comments to all the functions

jeffbuonamici commented on JavaFXUserInterface/src/main/java/external/ModelDAOImpl.java:
```diff
@@ -172,6 +184,64 @@ public String getGetModelEvaluation(int modelID, String assetTypeID) {
+        return modelStrategy;
+    }
+
+    @Override
+    public void updateModelStrategy(ModelStrategy modelStrategy, int modelID, int assetTypeID){
```
Add comments

jeffbuonamici commented on JavaFXUserInterface/src/main/java/external/ModelDAOImpl.java:
```diff
@@ -172,6 +184,64 @@ public String getGetModelEvaluation(int modelID, String assetTypeID) {
+        }
+    }
+
+    public double getLatestRMSE(int modelID,int assetTypeID){
```
Add comments

jeffbuonamici commented on Server/src/main/java/app/ModelController.java:
```diff
@@ -248,4 +261,16 @@ public Instances createInstancesFromAssets(List<Asset> assets) {
         return data;
     }
+
+    public void modelToEvaluate(TrainedModel trainedModel) throws Exception {
```
Add comments

rdsaliba replied:
@t98b There's multiple review by Jeff asking for comments. Did you add those? Look at the comments and resolve the conversations.

jeffbuonamici commented on Server/src/main/java/app/item/EvaluationList.java:
```diff
@@ -0,0 +1,18 @@
+public class EvaluationList {
+    public static ArrayList<EvaluateModel>  modelsList = new ArrayList<EvaluateModel>();
+
+//    private EvaluationList(){
+//        modelsList = new ArrayList<EvaluateModel>();
+//    }
+//
+//    private static EvaluationList instance;
+//
+//    public static EvaluationList getInstance() {
+//        if (instance == null) instance = new EvaluationList();
+//        return instance;
+//    }
```
Remove if no longer needed.

rdsaliba replied:
@t98b You still have this one!

jeffbuonamici commented on Server/src/main/java/local/ModelDAO.java:
```diff
@@ -19,6 +19,8 @@
     String getModelNameFromModelID(String modelID);
 
+    void updateRMSE(Double rmse, int modelId, int assetTypeId);
+    // TrainedModel createTrainedModelFromResultSet(ResultSet rs, boolean withModel) throws SQLException;
```
Remove if no longer needed.

jeffbuonamici commented on JavaFXUserInterface/src/main/java/Controllers/AssetTypeInfoController.java:
```diff
@@ -581,41 +528,64 @@ public void generateThumbnails() {
+                modelStrategy.setTrainAssets(from);
+                modelStrategy.setTestAssets(to);
+                modelDAO.updateModelStrategy(modelStrategy,1,1);
+                //                    EvaluateModel evaluateModel = new EvaluateModel();
+//                    evaluateModel.setAssetTypeID(assetTypeID);
+//                    evaluateModel.setModelName(modelName);
+//                    evaluateModel.setFrom(from);
+//                    evaluateModel.setTo(to);
+//                    modelList.add(evaluateModel);
+//                    modelDAO.insertToEvaluate(modelName, assetTypeID, (int) trainSlider.getValue(), (int) testSlider.getValue());
```
Remove if no longer needed.

jeffbuonamici commented on JavaFXUserInterface/src/main/java/Controllers/AssetTypeInfoController.java:
```diff
@@ -581,41 +528,64 @@ public void generateThumbnails() {
+                List<Asset> trainAssets= assets.subList(0,from-1);
+                List<Asset> testAssets= assets.subList(from,from+to-1);
+                local.ModelDAOImpl dao=new local.ModelDAOImpl();
+//                    ModelStrategy modelStrategy = modelDAO.getModelStrategy(model.getModelID(),assetTypeID);
```
Remove if no longer needed.

jeffbuonamici commented on JavaFXUserInterface/src/main/java/Controllers/AssetTypeInfoController.java:
```diff
@@ -284,125 +325,31 @@ public void enableEvaluation(List<Button> enableBtnList) {
-        double rmse = modelEvaluation.evaluateTrainWithTest();
-        modelDAO.updateRMSE(rmse, modelId, Integer.parseInt(assetType.getId()));
-    }
+//    private void evaluateAllModels() {
+//        setTrainAndTestInstances();
+//        try {
+//            evaluateModelClassifiers(modelObservableList);
+//        } catch (Exception exception) {
+//            exception.printStackTrace();
+//        }
+//    }
+
+//    /**
+//     * This function evaluates a selected model for the current asset type
+//     *
+//     * @author Talal, Jeremie
+//     */
+//    private void evaluateSelectedModel(Model selectedModel) {
+//        setTrainAndTestInstances();
+//        ObservableList<Model> selectedModelToEvaluateList = FXCollections.observableArrayList();
+//        selectedModelToEvaluateList.add(selectedModel);
+//        try {
+//            evaluateModelClassifiers(selectedModelToEvaluateList);
+//        } catch (Exception exception) {
+//            exception.printStackTrace();
+//        }
+//        selectedModelToEvaluateList.clear();
+//    }
```
Remove if no longer needed.

jeffbuonamici commented on JavaFXUserInterface/src/main/java/Controllers/AssetTypeInfoController.java:
```diff
@@ -244,17 +263,39 @@ private void attachEventsModelTab() {
+                        } catch (SQLException throwables) {
+                            throwables.printStackTrace();
+                        }
+//                        EvaluateModel evaluateModel = new EvaluateModel();
+//                        evaluateModel.setAssetTypeID(assetTypeID);
+//                        evaluateModel.setModelName(modelName);
+//                        evaluateModel.setFrom(from);
+//                        evaluateModel.setTo(to);
+//                        modelList.add(evaluateModel);
+//                        modelDAO.insertToEvaluate(modelName, assetTypeID, (int) trainSlider.getValue(), (int) testSlider.getValue());
```
Remove if no longer needed.

jeffbuonamici commented on JavaFXUserInterface/src/main/java/external/ModelDAO.java:
```diff
@@ -8,22 +8,31 @@
-    int getModelIDFromAssetTypeID(String assetTypeID);
+    TrainedModel createTrainedModelFromResultSet(ResultSet rs, boolean withModel) throws SQLException;
+    // int getModelIDFromAssetTypeID(String assetTypeID);
 
-    void updateRMSE(Double rmse, int modelId, int assetTypeId);
+    // void updateRMSE(Double rmse, int modelId, int assetTypeId);
```
Remove if no longer needed.

TalalBazerbachi committed:
#276 remove unsused function

TalalBazerbachi committed:
#276 add comments to functions

TalalBazerbachi commented:
all changes have been applied

najimgh commented on JavaFXUserInterface/src/main/java/external/ModelDAOImpl.java:
```diff
@@ -19,10 +26,11 @@
+    private static final String GET_LATEST_RMSE = "select rmse from trained_model where model_id=? and asset_type_id=? and status_id=2";
     private static final String UPDATE_MODEL_FOR_ASSET_TYPE = "UPDATE trained_model set model_id = ? where asset_type_id = ?";
     private static final String UPDATE_RETRAIN = "UPDATE trained_model SET retrain = true WHERE asset_type_id = ?";
+    private static final String GET_MODEL_STRATEGY = "select serialized_model from trained_model where model_id=? and asset_type_id=? and status_id=2";
```
It would be good to have SQL keywords in caps as it is our convention

najimgh commented on JavaFXUserInterface/src/main/java/external/ModelDAOImpl.java:
```diff
@@ -19,10 +26,11 @@
     private static final String GET_ALL_MODELS = "SELECT * from model";
-    private static final String GET_MODEL_EVALUATION = "SELECT rmse FROM model_evaluation WHERE model_id = ? AND asset_type_id = ?";
-    private static final String INSERT_RMSE = "REPLACE INTO model_evaluation SET rmse = ?,model_id = ?, asset_type_id = ? ";
+    private static final String GET_LATEST_RMSE = "select rmse from trained_model where model_id=? and asset_type_id=? and status_id=2";
```
It would be good to have SQL keywords in caps as it is our convention

TalalBazerbachi replied:
@najimgh I updated the queries

najimgh reviewed (changes requested):
You should add your changes to the database on top of the most recent cbms_script version (dumpfile v15 ?)

TalalBazerbachi committed:
#276 update sql queries to comply to project conventions

TalalBazerbachi commented:
@najimgh yes after you run cbms script 15, run script 16 (NOTE: if you ran v15 before, run it again)

TalalBazerbachi committed:
#276 create dump file for new changes on trained_model table

rdsaliba commented:
Hi @t98b. Everything seems fine to me. There's still the commented code as mentioned by Jeff thought.

TalalBazerbachi commented:
@rdsaliba 
can you please provide screenshots, so I know where there is still commented code?
Because I did remove all commented code. 
I might have missed some

rdsaliba commented:
<img width="830" alt="image" src="https://user-images.githubusercontent.com/17152773/111052741-add7e700-842b-11eb-86f2-b75b4ead1995.png">

<img width="848" alt="image" src="https://user-images.githubusercontent.com/17152773/111052751-ce07a600-842b-11eb-9930-d7c447cbaa88.png">

rdsaliba commented:
There's more. Look at Jeff comment's and resolve those. They indicate where in the code.

TalalBazerbachi committed:
#276 fix syntax error in sql query

TalalBazerbachi committed:
#276 remove unused code

TalalBazerbachi commented:
removed them

TalalBazerbachi committed:
#276 remove unused class

TalalBazerbachi commented:
I fixed all the changes requested

rdsaliba commented on JavaFXUserInterface/src/main/java/external/ModelDAO.java:
```diff
@@ -8,22 +8,28 @@
-    int getModelIDFromAssetTypeID(String assetTypeID);
-
-    void updateRMSE(Double rmse, int modelId, int assetTypeId);
+    TrainedModel createTrainedModelFromResultSet(ResultSet rs, boolean withModel) throws SQLException;
```
This method is never used. Why did you add it?

t98b replied:
getModelEvaluation?

TalalBazerbachi committed:
#276 remove reference of undeclared function

TalalBazerbachi committed:
#276 create dump file add rmse column in trained_model table

This reverts commit 46617998c416e9867c12c071b0c9febffd915cda.

TalalBazerbachi committed:
Merge branch 'develop' into 276_Move_Evaluate_To_Server

Khal-Shah reviewed (changes requested):
I ran V16 after V15 and ran server for several minutes but keep getting these errors: 
<img width="1268" alt="image" src="https://user-images.githubusercontent.com/35305423/111072616-b3781000-84b1-11eb-8e18-ab69b6be463f.png">

Also, the ZenHub for this issue is blank, could you put a brief description and acceptance criteria there for this issue and also reference the previous issue this fixes.

PaulMicu commented:
> I ran V16 after V15 and ran server for several minutes but keep getting these errors:
> <img alt="image" width="1268" src="https://user-images.githubusercontent.com/35305423/111072616-b3781000-84b1-11eb-8e18-ab69b6be463f.png">
> 
> Also, the ZenHub for this issue is blank, could you put a brief description and acceptance criteria there for this issue and also reference the previous issue this fixes.

this is a local issue, you need to go increase the max_allowed_packet on your local database

PaulMicu commented on Server/src/main/java/app/ModelController.java:
```diff
@@ -124,12 +123,31 @@ public void checkModels() {
                 .stream()
                 .filter(tm -> tm.getStatusID() == Constants.STATUS_LIVE)
                 .forEach(this::trainAndSave);
+        trainedModelsToRetrain
+                .stream()
+                .filter(tm -> tm.getStatusID() == Constants.STATUS_EVALUATION)
+                .findFirst().ifPresent(this::evaluate);
         trainedModelsToRetrain
                 .stream()
                 .filter(tm -> tm.getStatusID() == Constants.STATUS_EVALUATION)
                 .findFirst().ifPresent(this::trainAndSave);
     }
```
you cant do that....like i told you, simply add an if statement in trainAndSave after the model is trained to check if the statusID is Eval and then do the evaluate

TalalBazerbachi replied:
removed this code

PaulMicu commented on Server/src/main/java/app/ModelController.java:
```diff
@@ -124,12 +123,31 @@ public void checkModels() {
+        catch(Exception e){
+            e.printStackTrace();
+        }
+    }
```
you dont need this

TalalBazerbachi replied:
function removed

PaulMicu commented on Server/src/main/java/app/ModelController.java:
```diff
@@ -249,4 +265,21 @@ public Instances createInstancesFromAssets(List<Asset> assets) {
+     */
+    public void modelToEvaluate(TrainedModel trainedModel) throws Exception {
+        trainedModel.setAssetTypeID(1);
+        int trainAssets =trainedModel.getModelStrategy().getTrainsAssets();
+        int testAssets=trainedModel.getModelStrategy().getTestAssets();
```
its better to keep track of an ArrayList of assets instead of an int to avoid calling assetDaoImpl.getAssetsFromAssetTypeID(trainedModel.getAssetTypeID()) multiple times for no reason

TalalBazerbachi replied:
it doesn't work, you can't have a list in model strategy, it caused a nonTransient error, I tried to fix it for hours but it didn't work

JayChouteau replied:
I believe we fixed the transient error for the most part. However, the changing of int to an arraylist of assets will be postpone to a further issue due to time constraints.

PaulMicu commented on Server/src/main/java/app/item/Model.java:
```diff
@@ -1,12 +1,20 @@
     private String modelName;
     private int modelID;
     private String description;
+    private SimpleStringProperty rmse;
+
+    public Model() { rmse = new SimpleStringProperty();}
+
+    public SimpleStringProperty getRMSE(){ return rmse;}
+
+    public void setRMSE(String rmse){ this.rmse.setValue(rmse);}
```
you need to use TrainedModel to store the RMSE not Model

TalalBazerbachi replied:
I am using Model because Jeremie used Model to do his issue last sprint, I just added on top of what he did

JayChouteau replied:
The RMSE is now stored in the trained model table

PaulMicu commented on Server/src/main/java/app/ModelController.java:
```diff
@@ -219,21 +237,19 @@ public Double estimateRUL(Asset asset, Classifier classifier) {
      * @author Paul
      */
     public Instances createInstancesFromAssets(List<Asset> assets) {
+        System.out.println("in create");
```
remove this

TalalBazerbachi replied:
removed print statement

TalalBazerbachi committed:
#276 fix null pointer exception error

PaulMicu commented on JavaFXUserInterface/src/main/java/external/ModelDAOImpl.java:
```diff
@@ -106,6 +116,13 @@ public void updateRMSE(Double rmse, int modelId, int assetTypeId) {
                     newModel.setModelID(rs.getInt("model_id"));
                     newModel.setModelName(rs.getString("name"));
                     newModel.setDescription(rs.getString("description"));
+                    try (PreparedStatement pss = getConnection().prepareStatement(GET_LATEST_RMSE)) {
+                        pss.setInt(1, newModel.getModelID());
+                        pss.setString(2, String.valueOf(assetTypeID));
+                        try (ResultSet rss = pss.executeQuery()) {
+                            if (rss.next())
+                                newModel.setRMSE(rss.getString("rmse"));
+                        }}
```
Avoid doing 1 query per model, simply get the trainedmodels with status id 2 for the specified asset type id and then iterate over that list.. you will then only need 1 query instead of 9...

TalalBazerbachi replied:
changes applied

TalalBazerbachi committed:
#276 remove evaluate function

TalalBazerbachi committed:
#276 modify query to retreive rmse for all models

PaulMicu commented on JavaFXUserInterface/src/main/java/controllers/AssetTypeInfoController.java:
```diff
@@ -249,14 +264,33 @@ private void attachEventsModelTab() {
+
+
             try {
-                evaluateAllModelsBtn.setOnMouseClicked(mouseEvent -> evaluateAllModels());
+                evaluateAllModelsBtn.setOnMouseClicked(mouseEvent -> {
+                    for (Model model : modelObservableList){
+                        String modelName = model.getModelName();
+                        int assetTypeID = Integer.parseInt(assetType.getId());
+                        int trainAssets = (int) trainSlider.getValue() + 1;
+                        int testAssets = (int) trainSlider.getValue() + 1 + (int) testSlider.getValue();
+                        try {
+                            ModelStrategy modelStrategy = modelDAO.getModelStrategy(associatedModelID,assetTypeID);
+                            modelStrategy.setTrainAssets(trainAssets);
+                            modelStrategy.setTestAssets(testAssets);
+                            modelDAO.updateModelStrategy(modelStrategy,associatedModelID,assetTypeID);
+                        } catch (SQLException throwables) {
+                            throwables.printStackTrace();
+                        }
+                         }
+
+                });
```
This code is duplicated almost line by line on line 510, please reduce the duplication

TalalBazerbachi replied:
moved duplicated code to separate function

TalalBazerbachi committed:
#276 fix error in get_all_models query

TalalBazerbachi committed:
#276 remove unused imports, functions, and objects

TalalBazerbachi commented:
@PaulMicu unused objects and imports have been removed

TalalBazerbachi committed:
#276 removed unused objects from evaluation

TalalBazerbachi committed:
Merge branch 'develop' into 276_Move_Evaluate_To_Server

TalalBazerbachi committed:
#276 fix megre errors

TalalBazerbachi committed:
#276 fix merge error

TalalBazerbachi committed:
Merge branch 'develop' into 276_Move_Evaluate_To_Server

TalalBazerbachi committed:
#276 fix merge error

JayChouteau committed:
Merging from Origin + adding serializable (#276)
- Fixed Serializing model issue

JayChouteau committed:
Fixing evaluation display (#276)

JayChouteau committed:
Code Clean-up, comments, merge conflicts resolution (#276)
- addition of loggers
- comments on methods
- Fix a logic issue in calling the model to save evaluation in the asset type controller

JayChouteau committed:
Comments and Renaming to be more consistent (#276)

JayChouteau reviewed (dismissed):
Good work on moving the evaluation to server. It seems like the running the server and evaluating now works as intended. However, you need to let the server run for a while to train all models at least once in the first place.

Another comment, please make sure to run the code-clean-up, address warnings when you push code. Also, please be sure to comment where needed and give meaningful names to variables and functions when you create or edit them so that someone can understand what the code does by just reading the names.

TalalBazerbachi committed:
#276 fix error that caused null pointer exeption when evaluating multiple times

TalalBazerbachi committed:
#276 fix error

TalalBazerbachi committed:
#276 change status id to variable in queries

JayChouteau committed:
Query fixed (#276)

TalalBazerbachi committed:
#276 Add Alert when model strategy is null

JayChouteau committed:
Resolve Merge Conflicts (#276)

Khal-Shah commented:
The process for updating the RMSE is not very efficient or user friendly because of the decoupling of UI and server and how the a model has to be trained and saved in order for the RMSE to get saved to the DB. But the feature works as intended if the steps of having the serializable object in the DB and then running server after pressing on evaluate are followed correctly. Initially the RMSE was not aligned well with the evaluate button but @JayChouteau fixed that bug within this issue. Very good addition with the alert box as well. 

Approved.
